### PR TITLE
Enable tiered storage in AWS via IAM policy

### PR DESCRIPTION
**Pull request by vuldin**
This PR adds tiered storage feature in AWS deployments. Once merged, new PRs will be created for adding the same feature to other cloud providers. Tiered storage is disabled by default (`tiered_storage_enabled` in `vars.tf`). When tiered storage is enabled, then by default it is enabled across all topics in this PR.

### Steps to run
```
git clone https://github.com/redpanda-data/deployment-automation.git
cd deployment-automation
git checkout enable-si
cd aws
terraform init -upgrade
terraform apply -var "enable_monitoring=false" -var "tiered_storage_enabled=true"
cd ..
ansible-playbook --private-key `cat ~/.ssh/id_rsa.pub | awk '{print $2}'` -i hosts.ini -v ansible/playbooks/provision-node.yml -e '{ "redpanda": { "cluster": { "cloud_storage_segment_max_upload_interval_sec": 30 }}}'
```

Now connect via ssh to a broker, create a topic, produce some data, and check S3 bucket for remote segments. If you need to retrieve the IP address of one of the nodes, run `terraform output`.
```
ssh ubuntu@<redpanda-ec2-ip>
rpk topic create alog
BATCH=$(date) ; printf "$BATCH %s\n" {1..1000} | rpk topic produce alog
```

Run the last command 20+ times. View contents of the bucket in AWS console to see remote segments.

### CODE REVIEW

File: ansible/playbooks/start-redpanda.yml
Comment (vuldin) [inline ansible/playbooks/start-redpanda.yml:80]: These lines may not be needed. We may not want to enable global `cloud_storage_remote_read`/`cloud_storage_remote_write`, `cloud_storage_max_upload_interval_sec` is being set here only to make it easier to detect remote segments, and I was told `cloud_storage_credentials_source` was needed but I have seen this work without it (although I cannot reproduce).

**Comment by vuldin** [comment]
Question: Should shadow indexing (an enterprise feature) be enabled by default in this project? Right now it is.

File: ansible/playbooks/start-redpanda.yml
Comment (vuldin) [inline ansible/playbooks/start-redpanda.yml:89]: `cloud_storage_segment_max_upload_interval_sec` is being set here just to make it easy to see remote segments without waiting. It may need to be left out.

File: README.md
Comment (rkruze) [reply README.md:45]: I agree it is a bit verbose, but let's use the same name. I wonder at this point we might want to consider a config file with all the variables listed vs just being documented?

File: aws/cluster.tf
Comment (rkruze) [inline aws/cluster.tf:65]: Should this all be conditional based on having SI enabled?

File: aws/vars.tf
Comment (rkruze) [inline aws/vars.tf:109]: Let's call this tiered storage, not SI.

**Comment by rkruze** [comment]
Let's rename this to tiered storage instead of SI. As this is what the docs are focused around.

File: ansible/playbooks/start-redpanda.yml
Comment (r-vasquez) [inline ansible/playbooks/start-redpanda.yml:76]: Shouldn't this `access_key` and `secret_key` be overridden by an ansible var?

File: ansible/playbooks/start-redpanda.yml
Comment (r-vasquez) [inline ansible/playbooks/start-redpanda.yml:73]: I think we can add a var for api_urls that contains all the admin API endpoints, similar to what you did [here](https://github.com/redpanda-data/deployment-automation/commit/8579fdde423714696881c07dc1ee331652c31ffd#diff-c28b7b45769b3c8b11796f623bc3b3aa49a28fcd79d5f0cbb2d6ebc757994e8aR58)

File: aws/cluster.tf
Comment (tmgstevens) [reply aws/cluster.tf:65]: +1

File: aws/vars.tf
Comment (tmgstevens) [inline aws/vars.tf:18]: I'm not really sure about the purpose of `instance_name_prefix`. Perhaps we need a way of overriding the `deployment_id` so that it isn't always UUID-timestamp - so if a user wanted to put their project codename or something then they could.

File: ansible/playbooks/start-redpanda.yml
Comment (vuldin) [reply ansible/playbooks/start-redpanda.yml:76]: The only reason these two values are here is because they are required by Redpanda, but they are never being used in this implementation. This instead relies on the permissions available from the associated IAM profile.

But I agree that we would need to replace these values with user-provided ones if that approach was being implemented. If we do decide to implement that approach, then I was hoping for that to be in a separate PR so that this one just focuses on the IAM profile approach.

File: aws/vars.tf
Comment (vuldin) [reply aws/vars.tf:18]: I agree that these two variables are too similar. I'll change the name of this variable to `deployment_prefix` and have it be empty by default. Then I'll update the comment to know that that will override `deployment_id` if set, and that `deployment_id` already has a unique default value. I'll also switch all current uses of `instance_name_prefix` to `deployment.id`.

File: aws/cluster.tf
Comment (vuldin) [reply aws/cluster.tf:65]: I'll address this in the next push. The user will be able to set `deployment_prefix`, but by default it will be blank. If left blank, then the default `deployment_id` value will be used.

File: ansible/playbooks/start-redpanda.yml
Comment (tmgstevens) [reply ansible/playbooks/start-redpanda.yml:76]: I think we should set them to something like "THISVALUENOTUSED" and a comment nearby to explain why.

File: aws/cluster.tf
Comment (vuldin) [inline aws/cluster.tf:8]: Terraform complained about names being too long, so I shortened both the UUID and the timestamp.

File: aws/provider.tf
Comment (tmgstevens) [inline aws/provider.tf:5]: Please update the readme

**Comment by vuldin** [comment]
I'll squash the PR feedback commits once the PR is approved (let me know if you think this is unneeded).

File: ansible/playbooks/roles/redpanda_broker/templates/configs/tiered_storage.j2
Comment (vuldin) [inline ansible/playbooks/roles/redpanda_broker/templates/configs/tiered_storage.j2:5]: This makes all topics tiered-storage-enabled by default if the `tiered_storage_enabled` terraform variable is enabled.

File: ansible/playbooks/roles/redpanda_broker/vars/main.yml
Comment (vuldin) [inline ansible/playbooks/roles/redpanda_broker/vars/main.yml:8]: The `tiered_storage_bucket_name` ansible variable is pulled from `hosts.ini`, and is only defined if the `tiered_storage_enabled` terraform variable is `true`.

File: aws/cluster.tf
Comment (vuldin) [inline aws/cluster.tf:29]: This needs to be more limited to list only those needed permissions.

Comment (vuldin) [reply aws/cluster.tf:29]: Discussed this with the broader team. From that conversation:
> Having an policy that tightly limits redpanda to its own bucket is clearly the right thing to do, it's less obvious to me that restricting the verbs is helpful: it creates maintenance burden to keep those up to date as/when we change redpanda, and doesn't meaningfully change security if redpanda is the owner of the bucket.

So limiting this policy to the specific bucket (as shown in this file below) is the best approach, and avoids a maintenance burden as tiered storage capabilities are expanded.

File: ansible/playbooks/roles/redpanda_broker/templates/configs/defaults.j2
Comment (vuldin) [inline ansible/playbooks/roles/redpanda_broker/templates/configs/defaults.j2:13]: This resulted in an error due to the fact that `advertised_rpc_api` (unlike other listeners) isn't a list.

File: ansible/playbooks/roles/redpanda_broker/templates/configs/tiered_storage.j2
Comment (tmgstevens) [inline ansible/playbooks/roles/redpanda_broker/templates/configs/tiered_storage.j2:8]: Need to pass this in as a param from terraform otherwise we're assuming AWS only

**Comment by hcoyote** [comment]
Temporarily cherry picked the ansible2.14+ node_exporter fixes into my local repo in order to get this to run. fwiw, it seems to run ok with the ansible 7.1.0 bottle in homebrew (or, at least, it's behaving the same way so far as other clusters I've run this week from a cluster provisioning standpoint).

File: aws/cluster.tf
Comment (hcoyote) [reply aws/cluster.tf:8]: how safe is it to just grab the first 8 characters off the uuid? I haven't look at how go-uuid works to see if it's one of the time-based uuids.

File: aws/s3.tf
Comment (hcoyote) [inline aws/s3.tf:19]: Can we get a comment added on here on why we disable versioning (and maybe what happens, good or bad, if the user overrides this)?

**Review by hcoyote — APPROVED** [review]
added some questions for clarity, I don't see anything specific that should hold this back as is.

File: ansible/playbooks/roles/redpanda_broker/templates/configs/tiered_storage.j2
Comment (vuldin) [reply ansible/playbooks/roles/redpanda_broker/templates/configs/tiered_storage.j2:2]: No, this is just a placeholder. Unfortunately Redpanda requires this to be something even when we are using `aws_instance_metadata` (IAM permissions applied to the EC2 instance).

**Comment by tmgstevens** [comment]
Thanks for the review @hcoyote

**Pull request merged**
